File: DSA Practice/LeetCode/19. Remove Nth Node From End of List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int length = 0;
        ListNode headHolder = head;
        while(head != null){
            length++;
            head = head.next;
        }

        if (length == 1){
            return head;
        }

        head = headHolder;

        int target = length - n - 1;

        if (target == -1){
            return head.next;
        }

        length = 0; // repurposed to be second pointer
        while(head != null){
            if (length == target){
                try{
                    head.next = head.next.next;
                } catch (Exception e){
                    head.next = null;
                }
            }

            length++;
            head = head.next;
        }

        head = headHolder;
        return head;
    }
}

File: DSA Practice/LeetCode/19. Remove Nth Node From End of List/solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        var length : Int = 0
        var headHolder : ListNode? = head

        while(headHolder != null){
            length++
            headHolder = headHolder.next
        }

        if (length == 1){
            return headHolder
        }

        headHolder = head
        var target = length - n - 1

        if (target == -1){
            return headHolder?.next
        }

        length = 0
        while(headHolder != null){
            if (length == target){
                try{
                    headHolder.next = headHolder.next.next
                } catch(e : Exception){
                    headHolder.next = null
                }
            }

            length++
            headHolder = headHolder.next
        }

        return head
    }
}